AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStore/src/Structure.dfy

```diff
@@ -607,4 +719,82 @@ module {:options "/functionSyntax:4" } Structure {
          ToEncryptedHierarchicalKey(item, key.EncryptionContext[TABLE_FIELD]) == key
   {}
 
+  predicate MutationLockAttribute?(m: DDB.AttributeMap) {
+    && BRANCH_KEY_IDENTIFIER_FIELD in m && m[BRANCH_KEY_IDENTIFIER_FIELD].S?
+    && KEY_CREATE_TIME in m && m[KEY_CREATE_TIME].S?
+    && TYPE_FIELD in m && m[TYPE_FIELD].S?
+    && M_LOCK_UUID in m && m[M_LOCK_UUID].S?
+    && HIERARCHY_VERSION in m && m[HIERARCHY_VERSION].N? && m[HIERARCHY_VERSION].N == "1"
+
+    && 0 < |m[BRANCH_KEY_IDENTIFIER_FIELD].S|
+    && 0 < |m[TYPE_FIELD].S|
+    && 0 < |m[M_LOCK_UUID].S|
+
+    && (forall k <- m.Keys - {M_LOCK_ORIGINAL, M_LOCK_TERMINAL, HIERARCHY_VERSION} :: m[k].S?)
+
+    && m[TYPE_FIELD].S == MUTATION_LOCK_TYPE
+
+    && M_LOCK_ORIGINAL in m && m[M_LOCK_ORIGINAL].B? && 0 < |m[M_LOCK_ORIGINAL].B|
+    && M_LOCK_TERMINAL in m && m[M_LOCK_TERMINAL].B? && 0 < |m[M_LOCK_TERMINAL].B|
```

Comment:
since these are represented as bytes, should there be a tighter condition on what kind of bytes?

Reply:
This is a great question!
And is tightly defined in Mutation State Structures.

Reply:
I have added some clarifying comments.